File: android/src/edu/nus/comp/dotagridandroid/ui/renderers/GenericProgram.java
package edu.nus.comp.dotagridandroid.ui.renderers;

import static android.opengl.GLES20.*;

/* Generic Program
 * Default Shaders Compiling Routines
 */
public class GenericProgram implements Closeable {
	int programId, vs, fs;
	public GenericProgram (String vsSrc, String fsSrc) {
		vs = glCreateShader(GL_VERTEX_SHADER);
		glShaderSource(vs, vsSrc);
		glCompileShader(vs);
		String errStr = glGetShaderInfoLog(vs);
		if (errStr.length() > 0)
			System.out.println("Vertex Shader error:\n" + errStr);
		
		fs = glCreateShader(GL_FRAGMENT_SHADER);
		glShaderSource(fs, fsSrc);
		glCompileShader(fs);
		errStr = glGetShaderInfoLog(fs);
		if (errStr.length() > 0)
			System.out.println("Fragment Shader error:\n" + errStr);
		
		programId = glCreateProgram();
		glAttachShader(programId, vs);
		glAttachShader(programId, fs);
		glLinkProgram(programId);
		errStr = glGetProgramInfoLog(programId);
		if (errStr.length() > 0)
			System.out.println("Program Linking error:\n" + errStr);
	}
	
	public int getProgramId () {
		return programId;
	}

	public void close() {
		glUseProgram(0);
		glDeleteProgram(programId);
		glDeleteShader(vs);
		glDeleteShader(fs);
	}
}

File: DotAGridMobile-Fruity/DotAGridMobile/DotAGridMobile/GenericProgram.m
//
//  GenericProgram.m
//  DotAGridMobile
//
//  Created by apple on 11/6/14.
//  Copyright (c) 2014 NUS. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "GenericProgram.h"

@interface GenericProgram() {
    GLuint programId, vs, fs;
}
@end

@implementation GenericProgram
@synthesize programId = programId;

-(id) initWithVertexShaderSource:(const char *)vsSrc withFragmentShaderSource:(const char *)fsSrc {
    self = [super init];
    vs = glCreateShader(GL_VERTEX_SHADER);
    glShaderSource(vs, 1, &vsSrc, 0);
    glCompileShader(vs);
    char errString[65536];
    GLsizei errStringLen = 0;
    glGetShaderInfoLog(vs, 65535, &errStringLen, errString);
    if (errStringLen > 0)
        NSLog(@"Vertex Shader error:\n%@", [NSString stringWithUTF8String:errString]);
    fs = glCreateShader(GL_FRAGMENT_SHADER);
    glShaderSource(fs, 1, &fsSrc, 0);
    glCompileShader(fs);
    glGetShaderInfoLog(fs, 65535, &errStringLen, errString);
    if (errStringLen > 0)
        NSLog(@"Fragment Shader error:\n%@", [NSString stringWithUTF8String:errString]);
    programId = glCreateProgram();
    glAttachShader(programId, vs);
    glAttachShader(programId, fs);
    glLinkProgram(programId);
    glGetProgramInfoLog(programId, 65535, &errStringLen, errString);
    if (errStringLen > 0)
        NSLog(@"Program Link error:\n%@", [NSString stringWithUTF8String:errString]);
    return self;
}

-(void)close {
    glDeleteProgram(programId);
    glDeleteShader(vs);
    glDeleteShader(fs);
}
@end